### PR TITLE
Update Substrate

### PR DESCRIPTION
**Pull request by AurevoirXavier**

### CODE REVIEW

File: client/dvm/db/Cargo.toml [hunk 1]
```diff
@@ -12,13 +12,13 @@ version     = "2.7.0"
 [dependencies]
 # crates.io
 codec        = { package = "parity-scale-codec", version = "2.1", features = ["derive"] }
-kvdb         = { version = "0.9" }
-kvdb-rocksdb = { version = "0.11" }
+kvdb         = { version = "0.10" }
+kvdb-rocksdb = { version = "0.12" }
```
Comment (AsceticBear) [inline client/dvm/db/Cargo.toml:16]: https://github.com/paritytech/frontier/blame/master/client/db/Cargo.toml#L16-L17

File: client/dvm/rpc/src/eth.rs [hunk 2]
```diff
@@ -38,17 +38,20 @@ use sc_client_api::{
...
 	traits::{BlakeTwo256, Block as BlockT, NumberFor, One, Saturating, UniqueSaturatedInto, Zero},
 	transaction_validity::TransactionSource,
 };
 use sp_storage::{StorageData, StorageKey};
-use sp_transaction_pool::{InPoolTransaction, TransactionPool};
 // --- std ---
 use codec::{self, Decode, Encode};
-use ethereum::{Block as EthereumBlock, Transaction as EthereumTransaction};
+use ethereum::{
+	Block as EthereumBlock, LegacyTransaction, LegacyTransactionMessage, TransactionAction,
```
Comment (AsceticBear) [inline client/dvm/rpc/src/eth.rs:52]: 1. Recommend `BlockV0 as EthereumBlock` to replace `EthereumBlock<LegacyTransaction>`
2. `LegacyTransactionMessage` or `TransactionV0`. They are the same thing. Keep only one. Prefer the first.

**Comment by AsceticBear** [comment]
Why the CI stalled

File: node/service/src/chain_spec/pangoro.rs [hunk 3]
```diff
@@ -35,7 +34,8 @@ use darwinia_staking::StakerStatus;
 use drml_common_primitives::*;
 use pangoro_runtime::*;
 
-pub type ChainSpec = sc_service::GenericChainSpec<GenesisConfig>;
+// pub type ChainSpec = GenericChainSpec<GenesisConfig, Extensions>;
```
Comment (AsceticBear) [inline node/service/src/chain_spec/pangoro.rs:37]: unused comment?

File: node/service/src/chain_spec/pangolin.rs [hunk 4]
```diff
@@ -37,6 +36,7 @@ use darwinia_evm::GenesisAccount;
 use drml_common_primitives::*;
 use pangolin_runtime::*;
 
+// pub type ChainSpec = GenericChainSpec<GenesisConfig, Extensions>;
```
Comment (AsceticBear) [inline node/service/src/chain_spec/pangolin.rs:39]: unused comment?

File: node/service/src/chain_spec/mod.rs [hunk 5]
```diff
@@ -64,6 +71,21 @@ const TEAM_MEMBERS: &[&str] = &[
...
+// #[derive(Default, Clone, Serialize, Deserialize, ChainSpecExtension)]
+// #[serde(rename_all = "camelCase")]
+// pub struct Extensions {
+// 	/// Block numbers with known hashes.
+// 	pub fork_blocks: ForkBlocks<OpaqueBlock>,
+// 	/// Known bad block hashes.
+// 	pub bad_blocks: BadBlocks<OpaqueBlock>,
+// 	/// The light sync state extension used by the sync-state rpc.
+// 	pub light_sync_state: LightSyncStateExtension,
+// }
```
Comment (AsceticBear) [inline node/service/src/chain_spec/mod.rs:87]: useless comements?

File: node/service/src/chain_spec/pangoro.rs (hunk 3 above)
Comment (AurevoirXavier) [reply node/service/src/chain_spec/pangoro.rs:37]: It will be activated in the next version

Comment (AsceticBear) [reply node/service/src/chain_spec/pangoro.rs:37]: Better to add some comments?

Comment (AurevoirXavier) [reply node/service/src/chain_spec/pangoro.rs:37]: NP

File: client/dvm/rpc/src/eth.rs (hunk 2 above)
Comment (AurevoirXavier) [reply client/dvm/rpc/src/eth.rs]: TransactionV0 is corresponding to LegacyTransaction.

I think there is no alias for LegacyTransactionMessage.

File: frame/dvm/src/tests.rs [hunk 6]
```diff
@@ -173,59 +173,70 @@ fn get_wkton_balance(sender: &AccountInfo, nonce: u64) -> U256 {
...
 	let func = Function {
 		name: "balanceOf".to_owned(),
 		inputs: vec![Param {
 			name: "address".to_owned(),
 			kind: ParamType::Address,
+			internal_type: Some("address".into()),
 		}],
 		outputs: vec![],
 		constant: false,
+		state_mutability: StateMutability::NonPayable,
```
Comment (xiaoch05) [inline frame/dvm/src/tests.rs:186]: why the old version this `constant` is false?
the balanceOf function in wkton is a view function, here constant should be `true`, and `state_mutability=StateMutability::View`

**Review by xiaoch05 — CHANGES_REQUESTED** [review]

File: frame/dvm/src/tests.rs (hunk 6 above)
Comment (AsceticBear) [reply frame/dvm/src/tests.rs:186]: I think you are right! Update this  BTW? @AurevoirXavier

Comment (AurevoirXavier) [reply frame/dvm/src/tests.rs:186]: Not sure. Recommend you to take a full review after this.

**Review by AsceticBear — APPROVED** [review]

File: frame/dvm/rpc/runtime-api/src/lib.rs [hunk 7]
```diff
@@ -86,24 +86,24 @@ sp_api::decl_runtime_apis! {
 			estimate: bool,
 		) -> Result<dp_evm::CreateInfo, sp_runtime::DispatchError>;
 		/// Return the current block.
-		fn current_block() -> Option<EthereumBlock>;
+		fn current_block() -> Option<EthereumBlockV0>;
```
Comment (HackFisher) [inline frame/dvm/rpc/runtime-api/src/lib.rs:89]: Will types.json need to updated?

File: frame/staking/src/testing_utils.rs [hunk 8]
```diff
@@ -2,31 +2,32 @@
...
+// 		let validator_prefs = ValidatorPrefs {
+// 			commission: Perbill::from_percent(50),
+// 			..Default::default()
+// 		};
+// 		<Staking<T>>::validate(RawOrigin::Signed(controller).into(), validator_prefs)?;
+// 		let stash_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(stash);
+// 		validators.push(stash_lookup);
+// 	}
+// 	Ok(validators)
+// }
```
Comment (HackFisher) [inline frame/staking/src/testing_utils.rs:30]: Reason for comment out? Same with upstream?

File: node/runtime/pangolin/src/pallets/election_provider_multi_phase.rs [hunk 9]
```diff
@@ -13,6 +13,19 @@ sp_npos_elections::generate_solution_type!(
...
+impl pallet_election_provider_multi_phase::BenchmarkingConfig for BenchmarkConfig {
+	const VOTERS: [u32; 2] = [5_000, 10_000];
+	const TARGETS: [u32; 2] = [1_000, 2_000];
+	const ACTIVE_VOTERS: [u32; 2] = [1000, 4_000];
+	const DESIRED_TARGETS: [u32; 2] = [400, 800];
+	const SNAPSHOT_MAXIMUM_VOTERS: u32 = 25_000;
+	const MINER_MAXIMUM_VOTERS: u32 = 15_000;
+	const MAXIMUM_TARGETS: u32 = 2000;
+}
+
```
Comment (HackFisher) [inline node/runtime/pangolin/src/pallets/election_provider_multi_phase.rs:28]: Only used for benchmarks?

File: node/runtime/pangoro/src/pallets/babe.rs [hunk 10]
```diff
@@ -18,6 +18,7 @@ impl Config for Runtime {
 	type EpochDuration = EpochDuration;
 	type ExpectedBlockTime = ExpectedBlockTime;
 	type EpochChangeTrigger = ExternalTrigger;
+	type DisabledValidators = Session;
```
Comment (HackFisher) [inline node/runtime/pangoro/src/pallets/babe.rs:21]: Further review.

File: node/service/src/chain_spec/mod.rs (hunk 5 above)
Comment (HackFisher) [reply node/service/src/chain_spec/mod.rs:88]: Further review, what's the effect of disabling this?

File: node/service/src/service/pangolin.rs [hunk 11]
```diff
@@ -372,15 +371,11 @@ where
...
-			task_manager.spawn_handle(),
-			backend.clone(),
-			import_setup.1.shared_authority_set().clone(),
-		),
-	);
 
+	let warp_sync = Arc::new(sc_finality_grandpa::warp_proof::NetworkProvider::new(
+		backend.clone(),
+		import_setup.1.shared_authority_set().clone(),
+	));
```
Comment (HackFisher) [inline node/service/src/service/pangolin.rs:378]: Explanation and further review.

File: node/service/src/service/pangoro.rs [hunk 12]
```diff
@@ -297,15 +295,11 @@ where
...
 		.extra_sets
 		.push(sc_finality_grandpa::grandpa_peers_set_config());
-	config.network.request_response_protocols.push(
-		sc_finality_grandpa_warp_sync::request_response_config_for_chain(
-			&config,
-			task_manager.spawn_handle(),
-			backend.clone(),
-			import_setup.1.shared_authority_set().clone(),
-		),
-	);
```
Comment (HackFisher) [inline node/service/src/service/pangoro.rs:307]: Further review.

File: frame/dvm/rpc/runtime-api/src/lib.rs (hunk 7 above)
Comment (AurevoirXavier) [reply frame/dvm/rpc/runtime-api/src/lib.rs:89]: Not sure.
We had never maintained this.

File: frame/staking/src/testing_utils.rs (hunk 8 above)
Comment (AurevoirXavier) [reply frame/staking/src/testing_utils.rs:30]: Unused but may be helpful in the future.

File: node/runtime/pangolin/src/pallets/election_provider_multi_phase.rs (hunk 9 above)
Comment (AurevoirXavier) [reply node/runtime/pangolin/src/pallets/election_provider_multi_phase.rs:28]: Yes

File: node/service/src/chain_spec/mod.rs (hunk 5 above)
Comment (AurevoirXavier) [reply node/service/src/chain_spec/mod.rs:88]: Prepare for the next version.

File: node/service/src/service/pangoro.rs (hunk 12 above)
Comment (AurevoirXavier) [reply node/service/src/service/pangoro.rs:307]: paritytech/substrate#9227

File: node/service/src/service/pangolin.rs (hunk 11 above)
Comment (AurevoirXavier) [reply node/service/src/service/pangolin.rs:378]: paritytech/substrate#9227

File: node/runtime/pangoro/src/pallets/babe.rs (hunk 10 above)
Comment (AurevoirXavier) [reply node/runtime/pangoro/src/pallets/babe.rs:21]: paritytech/substrate#9414

**Review by HackFisher — APPROVED** [review]

File: frame/dvm/rpc/runtime-api/src/lib.rs (hunk 7 above)
Comment (HackFisher) [reply frame/dvm/rpc/runtime-api/src/lib.rs:89]: cc @sxlwar @chunqiuyiyu  , we might need to check this using Pangolin/Pangoro apps.

Comment (AsceticBear) [reply frame/dvm/rpc/runtime-api/src/lib.rs:89]: image

I tested this in my local env just now. It works well.

**Comment by AsceticBear** [comment]
Something to note: Make sure all the dvm-related tests pass before merging.

**Pull request merged**